src/Database/Query.php

```diff
@@ -369,6 +372,58 @@ public function traverseParts(callable $visitor, array $parts)
         return $this;
     }
 
+    /**
+     * Creates a common table expression (CTE).
+     *
+     * @param string $name The CTE name.
+     * @param \Cake\Database\ExpressionInterface|\Closure|string $query The CTE query definition.
+     * @return \Cake\Database\Expression\CommonTableExpression
+     */
+    public function commonTableExpression(string $name, $query): CommonTableExpression
+    {
+        if ($query instanceof Closure) {
+            $query = $query($this->newExpr(), $this);
+        }
+
+        return new CommonTableExpression($name, $query);
+    }
+
+    /**
+     * Adds a common table expression (CTE).
+     *
+     * @param \Cake\Database\Expression\CommonTableExpression|\Closure|null $expression The CTE to add.
+     * @param bool $overwrite Whether to reset the list of CTE's.
+     * @return $this
+     */
+    public function with($expression, $overwrite = false)
+    {
+        if ($overwrite) {
+            $this->_parts['with'] = null;
+        }
+
+        if ($expression === null) {
+            if (!$overwrite) {
+                throw new \InvalidArgumentException(
+                    'Resetting the WITH clause only works when overwriting is enabled.'
+                );
+            }
+
+            return $this;
+        }
+
+        if (!$this->_parts['with']) {
+            $this->_parts['with'] = new WithExpression();
```

Comment:
I don't think I understand why a separate `WithExpression` is needed. Why can't we just store an array of CommonTableExpressions?

Reply:
In the project where this stems from I had to test for recursive expressions quite a lot, and I didn't had to support multiple dialects. On project level I usually use concrete expressions for everything, as they're easy to test/reuse/extend/transform, extending the compiler on project level is always a little annoying.

Since the core needs compiler adjustments anyways, and given the fact that in the end it's just a stack of expressions, moving responsibilities into the compilers shouldn't be a problem.

Reply:
As discussed in the other PR, we could do something similar here:

`with('name', $cte)` with closure support.

That would move the CTE name to the query. I don't know if that would cause other issues.

Reply:
If there was only that signature, then pre-constructing CTEs would be rather awkward, either the name would need to be completely removed from the CTE, which feels weird given how essential the name is for a CTE, or the name would need to be made an optional property.

But pre-constructing CTEs without names has a weird vibe too, and from a syntactic point of view, the CTE query definition is required too, also it can refer to the name of the CTE, and refering to that name before it's even defined for the CTE itself, it's not a technical problem, but again feels rather weird.


```php
$cte2 = new CommonTableExpression($cteQuery1);
$cte2 = new CommonTableExpression(
    $connection->newQuery()->select('*')->from('cte1')->...
);
$cte3 = new CommonTableExpression(
    $connection->newQuery()
        ->select('*')
        ->from('cte2')
        ->unionAll(
            $connection->newQuery()
                ->select('*')
                ->from('cte3')
                ->...
        )
);

// ...

$query
    ->with('cte1', $cte1)
    ->with('cte2', $cte2)
    ->with('cte3', $cte3);
```

```php
$query->with('name', function (CommonTableExpression $cte, Query, $query) {
    return $cte
        ->setQuery($query->getConnection->newQuery()->...);
});
```

```php
$query->with('name', function (CommonTableExpression $cte, Query, $query) {
    return $query
        ->cte('name', $query->getConnection->newQuery()->...)
});
```

```php
$query->with('name', function (CommonTableExpression $cte, Query, $query) {
    return new CommonTableExpression(
        $query->getConnection->newQuery()->...
    );
});
```

`with()` could of course also be made to accept multipe signatures, like `with($cte, bool)` _and_ `with('name', $cte, bool)`, but I feel like this would make things just more confusing.

I have to think about this for a litte bit longer. I'll start with dropping `WithExpression`.

Reply:
If the name needs to be part of the cte then that's fine. But, once we decide on a signature for the closure, it's not going to change.

Reply:
I've tried various variants back and forth, and I'm afraid it just doesn't seem to fit like it does for the window expression, which has the advantage that it's syntactically valid even if it's empty, ie `OVER ()` and `win AS ()` are both valid.

I think if we wanted to go with passing an instance to the callable, then both the name and the query would need to be made optional and settable. Hover that deviates from other expressions, which do require the essential parts that are neccesary to build valid SQL, so it might not be overly appreciated. On the other hand, constructing nameless CTEs is weird, and having the name "optional" (by moving it out of the expression) but the query being required, is even weirder, and if the query would be made optional too, well then we're back to square one :)

Reply:
Let's not make it overly complicated/difficult to use. We can just allow setting the name outside the constructor and if you don't, it's an invalid expression.

Reply:
I tend to overthink things... Not sure if I understood you right, you mean like this?

```php
with(function (CommonTableExpression $cte, Query $query) {
    return $cte
        ->setName('cte1')
        ->setQuery(...);
})
```

Reply:
Yep. I was wrong earlier when I said `setName()` isn't needed because the name is required. This let's us use consistent conventions in places like the closure, I think.

Reply:
All right, let's go with that and see if there's any objections.